src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java

```diff
@@ -1093,28 +1026,23 @@ public TypeElement getFirstVisibleSuperClassAsTypeElement(TypeElement te) {
      *          be found.
      */
     public TypeMirror getFirstVisibleSuperClass(TypeMirror type) {
-        List<? extends TypeMirror> superTypes = typeUtils.directSupertypes(type);
-        TypeMirror superType = superTypes.isEmpty() ? getObjectType() : superTypes.get(0);
-        TypeElement superClass = asTypeElement(superType);
-        // skip "hidden" classes
-        while ((superClass != null && hasHiddenTag(superClass))
-                || (superClass != null &&  !isPublic(superClass) && !isLinkable(superClass))) {
-            TypeMirror supersuperType = superClass.getSuperclass();
-            TypeElement supersuperClass = asTypeElement(supersuperType);
-            if (supersuperClass == null
-                    || supersuperClass.getQualifiedName().equals(superClass.getQualifiedName())) {
-                break;
-            }
-            superType = supersuperType;
-            superClass = supersuperClass;
-        }
-        if (typeUtils.isSameType(type, superType)) {
-            return null;
+        // TODO: this computation should be eventually delegated to VisibleMemberTable
+        Set<TypeElement> alreadySeen = null;
+        assert (alreadySeen = new HashSet<>()) != null; // create set conditionally
```

Comment:
I think this use of `assert` is a step too far.  It's one thing to use asserts to verify invariants; it's too much to use them to conditionally compute state like this.

Reply:
That hash map is needed iff assertions are enabled. For an ordinary run, creating that hash map could be wasteful.

Perhaps I could create a hash map unconditionally using `HashSet.newHashSet(0)`; would that be to your liking?

Reply:
The issue was more about the composite declaration and use of the asserts to build a collection solely for use in the assertions. This is a level of assert above simply checking invariants with no other side effects on the JVM.  The use is clever and not wrong, but it is uncommon.  Maybe a higher level comment might help, more than the low level "create set conditionally"